internal/controller/apiextensions/composition/reconciler.go

```diff
@@ -156,34 +157,24 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 	}
 
 	var latestRev, existingRev int64
+
+	if lr := composite.LatestRevision(rl.Items); lr != nil {
+		latestRev = composite.LatestRevision(rl.Items).Spec.Revision
+	}
+
 	for i := range rl.Items {
 		rev := &rl.Items[i]
-		if !metav1.IsControlledBy(rev, comp) {
-			continue
-		}
-
-		if rev.Spec.Revision > latestRev {
-			latestRev = rev.Spec.Revision
-		}
 
-		want := v1alpha1.CompositionSpecDiffers()
-		if rev.GetLabels()[v1alpha1.LabelCompositionSpecHash] == currentHash {
+		if rev.GetLabels()[v1alpha1.LabelCompositionHash] == currentHash[:63] {
+			if rev.Spec.Revision != latestRev {
+				rev.Spec.Revision = latestRev + 1
+				if err := r.client.Update(ctx, rev); err != nil {
+					log.Debug(errUpdateRevSpec, "error", err)
+					r.record.Event(comp, event.Warning(reasonUpdateRev, err))
+					return reconcile.Result{}, errors.Wrap(err, errUpdateRevSpec)
+				}
+			}
 			existingRev = rev.Spec.Revision
-			want = v1alpha1.CompositionSpecMatches()
-		}
```

Comment:
We could avoid some nesting by continuing early here. e.g.:

```go
// This revision does not match our current Composition.
if rev.GetLabels()[v1alpha1.LabelCompositionHash] != currentHash[:63] {
	continue
}

// This revision matches our current Composition. We don't need a new one.
existingRev = rev.Spec.Revision

// This revision has the highest revision number - it doesn't need updating.
if rev.Spec.Revision == latestRev {
	continue
}

// This revision does not have the highest revision number. Update it so that it does.
rev.Spec.Revision = latestRev + 1
 if err := r.client.Update(ctx, rev); err != nil {
 	log.Debug(errUpdateRevSpec, "error", err)
 	r.record.Event(comp, event.Warning(reasonUpdateRev, err))
 	return reconcile.Result{}, errors.Wrap(err, errUpdateRevSpec)
 }
```

https://github.com/crossplane/crossplane/blob/master/CONTRIBUTING.md#return-early